src/Support/AutoDiscoveryHelper.php

```diff
@@ -2,115 +2,208 @@
 	 */
-	protected $filesystem;
+	protected $cache;
 	
 	/**
-	 * @var string
+	 * @var \Closure[] 
 	 */
-	protected $base_path;
+	protected $loaders;
 	
-	public function __construct(ModuleRegistry $module_registry, Filesystem $filesystem)
+	public function __construct(string $modules_path, CacheHelper $cache = null)
 	{
-		$this->module_registry = $module_registry;
-		$this->filesystem = $filesystem;
-		$this->base_path = $module_registry->getModulesPath();
+		$this->modules_path = rtrim($modules_path, DIRECTORY_SEPARATOR);
+		$this->cache = $cache;
+		
+		$this->loaders = $this->getLoaders();
 	}
 	
-	public function commandFileFinder(): FinderCollection
+	public function toArray()
 	{
-		if ($this->basePathMissing()) {
-			return FinderCollection::empty();
-		}
-		
-		return FinderCollection::forFiles()
-			->name('*.php')
-			->in($this->base_path.'/*/src/Console/Commands');
+		return Collection::make($this->loaders)
+			->map(function(Closure $loader) {
+				return $loader();
+			})
+			->toArray();
 	}
 	
-	public function factoryDirectoryFinder(): FinderCollection
+	public function modules() : ?Collection
 	{
-		if ($this->basePathMissing()) {
-			return FinderCollection::empty();
-		}
-		
-		return FinderCollection::forDirectories()
-			->depth(0)
-			->name('factories')
-			->in($this->base_path.'/*/database/');
+		return $this->load('modules');
 	}
 	
-	public function migrationDirectoryFinder(): FinderCollection
+	public function commands() : ?Collection
 	{
-		if ($this->basePathMissing()) {
-			return FinderCollection::empty();
-		}
-		
-		return FinderCollection::forDirectories()
-			->depth(0)
-			->name('migrations')
-			->in($this->base_path.'/*/database/');
+		return $this->load('commands');
 	}
 	
-	public function modelFileFinder(): FinderCollection
+	public function legacyFactoryPaths() : ?Collection
 	{
-		if ($this->basePathMissing()) {
-			return FinderCollection::empty();
-		}
-		
-		return FinderCollection::forFiles()
-			->name('*.php')
-			->in($this->base_path.'/*/src/Models');
+		return $this->load('legacy_factories');
 	}
 	
-	public function bladeComponentFileFinder() : FinderCollection
+	public function migrations() : ?Collection
 	{
-		if ($this->basePathMissing()) {
-			return FinderCollection::empty();
+		return $this->load('migrations');
+	}
+	
+	public function models(): ?Collection
+	{
+		return $this->load('models');
+	}
+	
+	public function bladeComponents() : ?Collection
+	{
+		return $this->load('blade_components');
+	}
+	
+	public function routes(): ?Collection
+	{
+		return $this->load('routes');
+	}
+	
+	public function viewDirectories() : ?Collection
+	{
+		return $this->load('view_directories');
+	}
+	
+	protected function load(string $name) : ?Collection
+	{
+		$result = $this->cache
+			? $this->cache->get($name)
+			: null;
+		
+		if (!$result) {
+			$result = $this->loaders[$name]();
+		}
+		
+		if ($result instanceof EmptyFinderCollection) {
+			return null;
 		}
```

Comment:
Not in love with this `EmptyFinderCollection` implementation, but it may be better than more conditionals up the stack from here.

Reply:
We might be able to put the `catch` in the `load()` method and then the whole thing can go away…